usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	ver, err := systemd.Version()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	// systemd 236 introduced the --collect option to systemd-run,
+	// which specifies that the unit should be garbage collected
+	// even if it fails.
+	//   https://github.com/systemd/systemd/pull/7314
+	if ver >= 236 {
+		args = append([]string{"systemd-run", "--user", "--collect", "--"}, args...)
+	} else {
+		args = append([]string{"systemd-run", "--user", "--"}, args...)
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+
+	if err := cmd.Run(); err != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q: %v", command, err))
+	}
+
+	return nil
+}
+
+var regularFileExists = osutil.RegularFileExists
+
+// desktopFileSearchPath returns the list of directories where desktop
+// files may be located.  It implements the lookup rules documented in
+// the XDG Base Directory specification.
+func desktopFileSearchPath() []string {
+	var desktopDirs []string
+
+	// First check $XDG_DATA_HOME, which defaults to $HOME/.local/share
+	dataHome := os.Getenv("XDG_DATA_HOME")
+	if dataHome == "" {
+		homeDir := os.Getenv("HOME")
+		if homeDir != "" {
+			dataHome = filepath.Join(homeDir, ".local/share")
+		}
+	}
+	if dataHome != "" {
+		desktopDirs = append(desktopDirs, filepath.Join(dataHome, "applications"))
+	}
+
+	// Next check $XDG_DATA_DIRS, with default from spec
+	dataDirs := os.Getenv("XDG_DATA_DIRS")
+	if dataDirs == "" {
+		dataDirs = "/usr/local/share/:/usr/share/"
+	}
+	for _, dir := range strings.Split(dataDirs, ":") {
+		if dir == "" {
+			continue
+		}
+		desktopDirs = append(desktopDirs, filepath.Join(dir, "applications"))
+	}
+
+	return desktopDirs
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+// Per https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id,
+// if desktop entries have dashes in the name ('-'), this could be an indication of subdirectories, so search
+// for those too. Eg, given foo-bar_baz_norf.desktop the following are searched for:
+//   o .../foo-bar_baz-norf.desktop
+//   o .../foo/bar_baz-norf.desktop
+//   o .../foo/bar_baz/norf.desktop
+//   o .../foo-bar_baz/norf.desktop
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(baseDir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(baseDir, strings.Join(splitFileId, "-"))
+
+	exists, isReg, _ := regularFileExists(desktopFile)
+	if exists && isReg {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (not all entries in $XDG_DATA_DIRS) and we know that snapd
+	// does not create subdirectories in that location.
```

Comment:
this comment doesn't sound valid anymore as the new approach seems to be to find the file as anything else would, and then double check that is from a place that is supported